include/seqan3/range/views/slice.hpp

```diff
@@ -49,21 +50,65 @@ struct slice_fn
                               std::ranges::range_difference_t<urng_t> begin_pos,
                               std::ranges::range_difference_t<urng_t> end_pos) const
     {
+        using position_t = std::ranges::range_difference_t<urng_t>;
         if constexpr (std::ranges::sized_range<urng_t>)
         {
-            using position_t = std::ranges::range_difference_t<urng_t>;
-            begin_pos = std::min(begin_pos, static_cast<position_t>(std::ranges::size(urange)));
-            end_pos = std::min(end_pos, static_cast<position_t>(std::ranges::size(urange)));
+            position_t urange_size = static_cast<position_t>(std::ranges::size(urange));
+
+            begin_pos = std::min(begin_pos, urange_size);
+            end_pos = std::min(end_pos, urange_size);
         }
+        position_t target_size = end_pos - begin_pos;
 
         if (end_pos < begin_pos)
             throw std::invalid_argument{"end_pos argument to seqan3::views::slice must be >= the begin_pos argument."};
 
-        // urange | drop | take
-        return views::take(views::drop(std::forward<urng_t>(urange), begin_pos), end_pos - begin_pos);
+        // SEQAN3_WORKAROUND_GCC_100139 == 1 if std::views::{take, drop} does not type reduce (e.g. keep in type
```

Comment:
Can you explain this to me again?
I would have thought that I need to use `views::type_reduce` if the `std::views` did not type erase; but apparently it's the other way around?

Reply:
Yeah, this is a bit complicated xD

so `std::views::drop(range_t, 2) -> view_t` has the following rules:

1. `range_t = std::span`, returns `view_t = std::span`
2. `range_t = std::string_view`, returns `view_t = std::string_view`
3. `range_t = std::ranges::subrange`, returns `view_t = std::ranges::subrange`

otherwise
* `view_t = std::ranges::drop_view<range_t>``

Currently, gcc 10 and 11 does always return `view_t = std::ranges::drop_view<range_t>`.

seqan3::views::type_reduce has some additional special rules:
1. `range_t is view`, returns view as is
  * this line is problematic if we do `view | type_reduce`,
  * as a `std::ranges::drop_view<range_t>` will be passed through;
  * that means only `type_reduce | view` makes sense
  * one could argue that this rule should be checked in a different order (at least that the borrowed_range ones have priority)
2. `range_t = std::string const &`, returns `view_t = std::string_view`
3. `range_t = std::ranges::contiguous_range && std::ranges::borrowed_range && std::ranges::sized_range`, returns `view_t = std::span`
4. `range_t = std::ranges::random_access_range && std::ranges::borrowed_range && std::ranges::sized_range`, returns `view_t = std::ranges::subrange`

----

Anyway since only `type_reduce | view` makes sense right now, we need that `std::views::drop` stays in type itself, otherwise it will always wrap `std::ranges::drop_view` around any type.

Example

view_t = range_t | seqan3::views::slice(0, 0);

| range_t | view_t if `SEQAN3_WORKAROUND_GCC_100139 == 1` | view_t if `SEQAN3_WORKAROUND_GCC_100139 == 0` | expected view_t |
| --- | --- | --- | --- |
| std::span | std::span | std::span \| type_reduce (1. Rule) \| drop (1. Rule) \| take (1. Rule) = std::span | std::span |
| std::string_view | std::string_view | std::string_view \| type_reduce (1. Rule) \| drop (2. Rule) \| take (2. Rule) = std::string_view | std::string_view |
| std::ranges::subrange | std::ranges::subrange | std::ranges::subrange \| type_reduce (4. Rule?) \| drop (3. Rule) \| take (3. Rule) = std::ranges::subrange | std::ranges::subrange |
| std::string const & | std::string_view | std::string const & \| type_reduce (2. Rule) \| drop (2. Rule) \| take (2. Rule) = std::string_view | std::string_view |
| std::ranges::contiguous_range && std::ranges::borrowed_range && std::ranges::sized_range | std::span | range_t \| type_reduce (3. Rule) \| drop (1. Rule) \| take (1. Rule) = std::span | std::span |
| std::ranges::random_access_range && std::ranges::borrowed_range && std::ranges::sized_range | std::ranges::subrange | range_t \| type_reduce (4. Rule) \| drop (3. Rule) \| take (3. Rule) = std::ranges::subrange | std::ranges::subrange |